src/libmongoc/src/mongoc/mongoc-cursor.c

```diff
@@ -1449,9 +1469,16 @@ void
 mongoc_cursor_set_batch_size (mongoc_cursor_t *cursor, uint32_t batch_size)
 {
    BSON_ASSERT (cursor);
+   bson_iter_t iter;
 
-   _mongoc_cursor_set_opt_int64 (
-      cursor, MONGOC_CURSOR_BATCH_SIZE, (int64_t) batch_size);
+   if (bson_iter_init_find (&iter, &cursor->opts, MONGOC_CURSOR_BATCH_SIZE) &&
+       BSON_ITER_HOLDS_INT32 (&iter)) {
+      _mongoc_cursor_set_opt_int32 (
+         cursor, MONGOC_CURSOR_BATCH_SIZE, (int32_t) batch_size);
+   } else {
+      _mongoc_cursor_set_opt_int64 (
+         cursor, MONGOC_CURSOR_BATCH_SIZE, (int64_t) batch_size);
+   }
```

Comment:
I had not considered other numeric types when filing the ticket. The server code appears to specify the `batchSize` option in the IDL as a `safeInt64`: https://github.com/10gen/mongo/blob/3aa51a4623b5869c6f57745ad1371aacf02c8d34/src/mongo/db/query/find_command.idl#L124-L129

A `safeInt64` appears to be representable as `int32`, `int64`, `double`, or `decimal128`: https://github.com/10gen/mongo/blob/3aa51a4623b5869c6f57745ad1371aacf02c8d34/src/mongo/db/basic_types.idl#L66-L74:

```
    safeInt64:
        bson_serialization_type:
                                - long
                                - int
                                - decimal
                                - double
        description: "Accepts any numerical type within int64 range"
        cpp_type: std::int64_t
        deserializer: "mongo::BSONElement::safeNumberLong"
```

I suggest also handling `double` and `decimal` similarly.